### ISSUE TITLE
madduck repo seems abandoned

### ISSUE THREAD

**Issue by sivann** [saltstack/salt-bootstrap#65]
madduck repo, used for debian, is still on 12.1. 
Do we have any repo alternatives? Should I report this somewhere else?

**Comment by s0undt3ch** [comment]
@madduck can I bring your attention to this ticket?

**Comment by madduck** [comment]
I only backport what's in Debian unstable. The problem is that there aren't really any Salt maintainers for Debian right now. The people who have been doing it are overloaded (as am I). We need someone to step up and just do it.

I have been travelling the last couple of weeks, which is why I have also not been very responsive. Hopefully I can work through my backlog and get to Salt issues soon…

**Comment by s0undt3ch** [comment]
@madduck Thanks for your input.

Yes, we need to find a way to automate this process and remove pressure from contributors like you, which can contribute, but not as often as sometimes needed because, of course, everyone has it's own priorities.

**Comment by sivann** [comment]
I can also offer a script that creates the libzmq3 and python-zmq packages for debian 6 from source, to avoid using easy_install on the bootstrap.

**Comment by benallard** [comment]
Can SaltStack provide some kind of hosting ? This way, we could setup a buildbot farm that automatically builds all the packages upon some kind of trigger ... @saltstack not having this kind of stuff is like the children of the shoemakers going barefoot ...

**Comment by s0undt3ch** [comment]
@benallard having such a build farm is already on our TODO list, yet, ASAP...

**Comment by umeboshi2** [comment]
I have packages built for wheezy, however I'm using fresh git version, instead of latest release, since I needed to report an issue recently.  The difference in these packages is that I am using zeromq 3.2.2 and pyzmq 13.0.0, as well as salt 0.13.2~ .  The debian tree is based on the source package in sid.  There are several problems with using the earlier 2.2 versions of zeromq and salt together, and to use these packages, multiprocessing must be set to False on the minions.

I feel that backporting this to squeeze should be fairly straightforward, although time consuming.  The builddep on the latest cython makes the backporting something that I would have to do over a weekend.  If there is enough interest in having these built for squeeze, let me know and I'll build them this weekend (time permitting).

Edit:  Packages are here:  http://debrepos.littledebian.org/paella

**Comment by s0undt3ch** [comment]
@umeboshi2 are you offering your time(time permitting, of course) and server to host the latest salt stable versions along with pyzmq compiled against ZMQ3 until we get the required infrastructure ready for @saltstack to build and host it's own packages?

**Comment by umeboshi2** [comment]
I'm offering to do the build.  It's going to be a bit more involved for squeeze, as I will have to backport debhelper, and put that in a separate builddep repository to keep it (and possibly cython) from interfering with the rest of a normal squeeze release.

As far as hosting, I'm willing to do so as long as it doesn't become costly.  If the bandwidth use gets too high, it will need to be mirrored.

**Comment by sivann** [comment]
well, I will include here instructions on how to build zmq3 and python-zmq packages for squeeze using [fpm](https://github.com/jordansissel/fpm/wiki) , since the debian-provided libzmq1 is pretty worthless. It may be useful to someone:

```
mkdir /src/slash/  #empty dir
/src/zeromq-3.2.2# ./configure --prefix=/src/slash/usr ; make ; make install
fpm -s dir -t deb -n libzmq3 -v 3.2.2 \
  --conflicts "libzmq1" \
  --replaces "libzmq1" \
  --replaces "libzmq3 (<< 3.2.2)" \
  --url "http://www.zeromq.org/" \
  --vendor "" --maintainer "user@company.com" \
  --description "lightweight messaging kernel" \
  -C /src/slash/ usr/
```

and for python-zmq:

```
rm -fr /src/slash/*
mkdir /src/slash/usr/
cd /src/pyzmq-2.2.0.1
./setup.py build
./setup.py install --prefix=/src/slash/usr --install-layout=deb  
cd ..
fpm -s dir -t deb \
  -n python-zmq -v 2.2.0.1 \
  --conflicts "python-zmq" \
  --depends "libzmq3" \
  --replaces "python-zmq" \
  --url "http://www.zeromq.org/bindings:python"  \
  --provides "python-zmq"   \
  --description "0MQ is a fast messaging library, linked with libzmq3"   \
  --iteration 3 \
  --vendor "" --maintainer "user@company.com"  \
  -C /src/slash/ usr/
```

Since writing this pyzmq jumped to version 13 but the rest is the same.

**Comment by madduck** [comment]
@umeboshi2 why do you need to backport debhelper? The packaging in Debian unstable works fine and can be trivially backported.

**Comment by umeboshi2** [comment]
Lol!  I just sent the zmq dscfile to the builder and pbuilder spit out depends problem and debhelper was at top of list.  I didn't even bother to look at it as I was just trying to guess, comment, and get back to work.  Sorry about that.

I turned out to be missing libpgm-dev.

I have often before had to backport debhelper to build a package more easily than messing with the code otherwise, and I just assumed this would be the case here as well.

**Comment by sivann** [comment]
Please when building, also have in mind that tags in debian/changelog are usually wrong: https://github.com/saltstack/salt/issues/3918

**Comment by ghost** [comment]
backporting debhelper is madness. that's nothing to laugh at.  construct your packages to use the native debhelper version and associated hooks for the target system. it's OK to use debhelper from the distro's official backports/unstable/etc repos as @madduck points out. This is why we use what looks like old stuff in salt's debian directory: because we build on platforms as old as Ubuntu Lucid (10.04) and as new as Raring Ringtail, so you have to use what works evenly on all the platforms that share a common packaging directory.

@sivann: though it does get out of sync sometimes (and doesn't really matter much), generally we try to keep debian/changelog with a generic, current native version

**Comment by umeboshi2** [comment]
I don't need to backport debhelper, just assumed that by glancing at pbuilder's output.  What was really needed was libpgm, which isn't in squeeze.  I already backported that.  libzmq is also currently backported to squeeze.  The nontrivial part of all of this is pyzmq.  It requires cython 0.18.  My wheezy backport of this went flawlessly because the python3 packages on wheezy are more up to date.

I am currently trying to build cython for squeeze without python3.  I'll probably also have to modify pyzmq to only build python2 packages.

I haven't even bothered building salt packages.  On my littledebian repository, I built a fresh version from git so that I could help report the issues I was experiencing easier.  I don't want to do this on the official salt repository.

However, I would like to get some advice on how to use the version names so that upgrades from squeeze to wheezy will work ok.  Debian has a useful, albeit an intricate and sometimes confusing, versioning system.  I sometimes have problems remembering how to properly version things so that later revisions will install, etc.

The actual salt debian repository is up and running.  I put salt's depends for wheezy in the repo, and I am buildling for squeeze now.  0mq and libpgm are already in the repo.  The host is going to be named by UtahDave in the near future.  Probably something like debian.saltstack.com or similar.

**Comment by ghost** [comment]
I can build salt packages on debian from the existing debian/ directory and would advise you to use that so we have a known common base to work from for officially supported packages. I welcome any adaptations there to make it work better & on more distros.

There are similar dependency issues for ubuntu and we also provide those in the PPA. most of them probably work unmodified on debian also, we've just never had a site to host on. You can download the package sources from the PPA by clicking on 'show package details' and expanding the triangle next to anything you're interested in and see what's in that package's debian/ dir.

agreed re: versioning. I just try to keep it consistent and use another PPA to stage/test things to see if they supersede before publishing them in the main PPA. https://launchpad.net/~saltstack/+archive/salt

HTH, @umeboshi2 !

**Comment by s0undt3ch** [comment]
@umeboshi2 You've been working on the Saltstack Debian repo right?
Let me know when you're finished, or please submit a pull request with the required changes to the bootstrap script.
I was unable to find the GnuPG key in the repository, you haven't uploaded it yet right?

**Comment by umeboshi2** [comment]
I uploaded it to a keyserver.

gpg --recv-keys EF031D06

**Comment by s0undt3ch** [comment]
Ok, I'll see if update the script then. Thanks!
Oh, the repository is already usable, or you still need to some more stuff before we can start hammering the repo?

**Comment by umeboshi2** [comment]
Only the depends for salt are on the repository.  For squeeze, I still need to build pyzmq.  For wheezy, everything but salt is ready to go.  As far as salt is concerned, I was going to concentrate on pyzmq for squeeze first, then start setting the debian tree in the repository to use quilt, so that this tree can be prepatched to build for particular releases. (Example apply squeeze patch, build for squeeze.  Apply ubuntu qq patch, build for qq and rr.)

**Comment by s0undt3ch** [comment]
Ok. Let' me know when I can make the changes to the script.

**Comment by sivann** [comment]
@seanchannel about git tagging getting out of sync with changelog: it matters because if this happens the new debian packages created with the new tag will have the same minor version with the previous tag and thus they will not be upgraded with apt.

**Comment by joehealy** [comment]
Hey umeboshi2!, where are you up to with this?

Can I give you a hand?

Cheers,

Joe

**Comment by umeboshi2** [comment]
madduck suggested that I build a zeromq3 package, which will sit alongside the zeromq package.  I built the zeromq3 package on my littledebian.org repo, but I am having problems with pyzmq.  I have also been too busy last week and I have a decent sized job on the horizon that will take time away.  I would happily welcome any help or suggestions.

I am happy with the state of things now on the saltstack debian machine.  I am under the impression that my zeromq (v3.2) will work just as well with packages that need zeromq2, as well as the pyzmq that I have built, and I have not really found a truly compelling reason to create zeromq3, excepting madduck's generous offer to help upload the packages to official debian repo.

**Comment by joehealy** [comment]
Maybe it is just me, but I can't see any salt packages there.

Am I looking in the wrong places?

**Comment by benallard** [comment]
They are here: http://debrepos.littledebian.org/paella/pool/main/s/salt/

**Comment by ghost** [comment]
For building libzmq3 and/or pyzmq3 you may be able to recycle the ones
from the Ubuntu PPA on Debian (get there -> package details -> little
triangle next to the package -> download tar.gz).  In particular look at
the 'lucid' versions, I suggest. These just might work as-is, but could
probably be shined up or updated to use newer debhelper, etc.

FWIW,
_S.

**Comment by joehealy** [comment]
Is there a reason they are not at debian.saltstack.com?

That seems to be the logical place for them.

**Comment by umeboshi2** [comment]
The salt packages on littledebian are built from an arbitrary revision of the git repo.  At the time, the version number was subsequent to the current stable, although this is irrelevant today.  Also, I used the debian tree from the official debian sid package (more or less) instead of the debian tree already in salt.  I didn't want to build salt packages on debian.saltstack.com in the manner I built my own packages.

Instead of just focusing on the debian tree, I've been sidetracked a bit by the suggestion to build a zeromq3 package.  I'm still not entirely convinced about the need for a zeromq3 package, and I am happy with the packages that I have already built that exist on the debian.saltstack.com site.

If there aren't any concerns about how the currently built zeromq and pyzmq packages perform on debian/ubuntu systems, then the main stack is already there and we can talk about the range of debian/ubuntu releases to attempt to support and create the debian tree with quilt to make the job easier.
